### PR TITLE
Add WindNet and Site PV

### LINKED ISSUE #123: India Wind prototype
Use PVnet to create a first forecast for wind. 

The deployment step is in https://github.com/openclimatefix/india-forecast-app/issues/11

### LINKED ISSUE #124: India PV prototype

### LINKED ISSUE #126: Adapting PVnet for Sites

### PR DESCRIPTION
**Pull request by jacobbieker**
# Pull Request

## Description

Adds the WindNet model

Fixes #123 
Fixes #124 
Fixes #126 

## How Has This Been Tested?

Unit Tests

- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: scripts/save_batches.py [hunk 1]
```diff
@@ -112,13 +116,13 @@ def main(config: DictConfig):
         batch_size=None,  # batched in datapipe step
         sampler=None,
         batch_sampler=None,
-        num_workers=config_dm.num_workers,
+        num_workers=0,  # config_dm.num_workers,
```
Comment (dfulu) [inline scripts/save_batches.py]: Can we put these back to be configurable

File: scripts/save_batches.py [hunk 2]
```diff
@@ -70,9 +74,9 @@ def _get_datapipe(config_path, start_time, end_time, batch_size, renewable: str
         end_time=end_time,
     )
 
-    data_pipeline = (
-        data_pipeline.batch(batch_size).map(stack_np_examples_into_batch).map(batch_to_tensor)
-    )
+    # data_pipeline = (
```
Comment (dfulu) [inline scripts/save_batches.py]: Can we keep this in in some way. Maybe if `renewable == "pv"` or `output_format == "torch"` or something

File: scripts/save_batches.py [hunk 3]
```diff
@@ -20,21 +20,23 @@
 import shutil
 import sys
 
+import dask
+
+dask.config.set(scheduler="single-threaded")
```
Comment (dfulu) [inline scripts/save_batches.py]: Is this needed? Could it instead go in the relevant part of the datapipe where the values are computed.  Like using the `.compute(scheduler="single-threaded")`

File: pvnet/models/utils.py [hunk 4]
```diff
@@ -110,7 +118,7 @@ def flush(self) -> dict[BatchKey, list[torch.Tensor]]:
         """Concatenate all accumulated batches, return, and clear self"""
         batch = {}
         for k, v in self._batches.items():
-            if k == BatchKey.gsp_t0_idx:
+            if k == BatchKey.gsp_t0_idx or k == BatchKey.wind_t0_idx or k == BatchKey.pv_t0_idx:
```
Comment (dfulu) [inline pvnet/models/utils.py:121]: can we use `if k in [BatchKey.gsp_t0_idx, BatchKey.wind_t0_idx, BatchKey.pv_t0_idx]` to be a little cleaner

File: pvnet/models/multimodal/site_encoders/encoders.py [hunk 5]
```diff
@@ -244,3 +244,295 @@ def forward(self, x):
         x_out = attn_output.squeeze()
 
         return x_out
+
+
+class SingleWindAttentionNetwork(AbstractPVSitesEncoder):
```
Comment (dfulu) [inline pvnet/models/multimodal/site_encoders/encoders.py:249]: It seems like there is probably quite a lot of duplication here. Could we do some inheritance to clean things up?

File: pvnet/models/multimodal/multimodal.py [hunk 6]
```diff
@@ -176,12 +186,21 @@ def __init__(
             assert pv_history_minutes is not None
 
             self.pv_encoder = pv_encoder(
-                sequence_length=pv_history_minutes // 5 + 1,
+                sequence_length=pv_history_minutes // 15,
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py]: Can we make this configurable so it will still work on the UK data

File: pvnet/models/multimodal/multimodal.py [hunk 7]
```diff
@@ -218,7 +237,7 @@ def forward(self, x):
             sat_data = x[BatchKey.satellite_actual][:, : self.sat_sequence_len]
             sat_data = torch.swapaxes(sat_data, 1, 2).float()  # switch time and channels
             if self.add_image_embedding_channel:
-                id = x[BatchKey.gsp_id][:, 0].int()
+                id = x[BatchKey.wind_id][:, 0].int()
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py]: Can we make this configurable?

File: pvnet/models/multimodal/multimodal.py [hunk 8]
```diff
@@ -253,10 +274,27 @@ def forward(self, x):
 
         # ********************** Embedding of GSP ID ********************
         if self.embedding_dim:
-            id = x[BatchKey.gsp_id][:, 0].int()
+            if self.target_key_name == "wind":
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py:300]: Would it be possible to clean with up with something like `id = x[BatchKey[f"{self.target_key_name}_id"]][:, 0].int()`? Not so sure about this one though

File: pvnet/models/multimodal/multimodal.py [hunk 9]
```diff
@@ -96,23 +100,28 @@ def __init__(
...
             target_key: The key of the target variable in the batch.
+            interval_minutes: The interval between each sample of the target data
         """
 
         self.include_gsp_yield_history = include_gsp_yield_history
         self.include_sat = sat_encoder is not None
         self.include_nwp = nwp_encoders_dict is not None and len(nwp_encoders_dict) != 0
         self.include_pv = pv_encoder is not None
         self.include_sun = include_sun
+        self.include_gsp = include_gsp
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py:122]: Is this used?

File: .gitignore [hunk 10]
```diff
@@ -140,3 +140,4 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+!/configs/datamodule/configuration/leonardo_wind_configuration.yaml
```
Comment (dfulu) [inline .gitignore]: As below, if we get rid of the configs dir, could this go?

File: pvnet/data/wind_datamodule.py [hunk 11]
```diff
@@ -1,9 +1,8 @@
 """ Data module for pytorch lightning """
```
Comment (dfulu) [inline pvnet/data/wind_datamodule.py:1]: There are some things in this file that were commented out but for cleanliness should probably just be deleted

File: pvnet/models/utils.py (hunk 4 above)
Comment (dfulu) [reply pvnet/models/utils.py:121]: Or even `k==f"{self.key_to_keep}_t0_idx"`

**Review by dfulu — CHANGES_REQUESTED** [review]

File: pvnet/data/pv_site_datamodule.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
+""" Data module for pytorch lightning """
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply pvnet/data/pv_site_datamodule.py:1]: Yeah, that is probably the right way to do it, I'll take a crack at it

File: pvnet/models/multimodal/encoders/encoders3d.py [hunk 13]
```diff
@@ -38,6 +38,7 @@ def __init__(
 
         # Check that the output shape of the convolutional layers will be at least 1x1
         cnn_spatial_output_size = image_size_pixels - 2 * number_of_conv3d_layers
+        cnn_sequence_length = sequence_length  # - 6 * number_of_conv3d_layers
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply pvnet/models/multimodal/encoders/encoders3d.py:41]: Yep

[the remark this answers is not in the document]
Comment (jacobbieker) [reply pvnet/models/multimodal/encoders/encoders3d.py:41]: Although I was thinking of changing this a bit to allow for customizable kernel sizes, so it won't match sequence length exactly. I'll make those changes

File: pvnet/models/base_model.py [hunk 14]
```diff
@@ -260,13 +267,15 @@ def __init__(
         self.output_quantiles = output_quantiles
 
         # Number of timestemps for 30 minutely data
-        self.history_len_30 = history_minutes // 30
-        self.forecast_len_30 = forecast_minutes // 30
+        self.history_len_30 = history_minutes // interval_minutes
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply pvnet/models/base_model.py]: Sounds good!

File: pvnet/models/base_model.py [hunk 15]
```diff
@@ -421,11 +432,20 @@ def _training_accumulate_log(self, batch, batch_idx, losses, y_hat):
...
+                    quantiles=self.output_quantiles,
+                    key_to_plot=self._target_key_name,
+                )
                 fig.savefig("latest_logged_train_batch.png")
+                plt.close(fig)
 
     def training_step(self, batch, batch_idx):
         """Run training step"""
+        # Make all -1 values 0.0
+        batch[self._target_key] = batch[self._target_key].clamp(min=0.0)
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply pvnet/models/base_model.py:456]: Yeah, that makes sense

File: pvnet/training.py [hunk 16]
```diff
@@ -105,9 +105,7 @@ def train(config: DictConfig) -> Optional[float]:
             if isinstance(callback, ModelCheckpoint):
                 # Need to call the .experiment property to initialise the logger
                 wandb_logger.experiment
-                callback.dirpath = "/".join(
-                    callback.dirpath.split("/")[:-1] + [wandb_logger.version]
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply pvnet/training.py:109]: For me, it was throwing errors with wandb_logger, even when the wandb_logger was initialized. I'll look back into it and re-add it

File: scripts/save_batches.py (hunk 2 above)
Comment (jacobbieker) [reply scripts/save_batches.py]: Yep!

**Review by peterdudfield — APPROVED** [review]
Looks great, merge away once the tests pass

File: pvnet/models/base_model.py [hunk 17]
```diff
@@ -372,15 +389,13 @@ def _calculate_val_losses(self, y, y_hat):
 
             # Take median value for remaining metric calculations
             y_hat = self._quantiles_to_prediction(y_hat)
+        # common_metrics_each_step = common_metrics(predictions=y_hat.numpy(), target=y.numpy())
+        mse_each_step = torch.sqrt(torch.mean((y_hat - y) ** 2, dim=0)) ** 2
```
Comment (dfulu) [inline pvnet/models/base_model.py:393]: We should simplify this to `torch.mean((y_hat - y) ** 2, dim=0)`

File: pvnet/models/multimodal/multimodal.py [hunk 18]
```diff
@@ -124,15 +152,17 @@ def __init__(
...
+            ) // sat_interval_minutes + 1
 
             self.sat_encoder = sat_encoder(
                 sequence_length=self.sat_sequence_len,
                 in_channels=sat_encoder.keywords["in_channels"] + add_image_embedding_channel,
             )
             if add_image_embedding_channel:
                 self.sat_embed = ImageEmbedding(
-                    318, self.sat_sequence_len, self.sat_encoder.image_size_pixels
+                    image_embedding_dim, self.sat_sequence_len, self.sat_encoder.image_size_pixels
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py:165]: Can we rename this to something like `num_embeddings` which would be in keeping with pytorch https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html

File: pvnet/models/multimodal/multimodal.py [hunk 19]
```diff
@@ -166,7 +197,9 @@ def __init__(
                 )
                 if add_image_embedding_channel:
                     self.nwp_embed_dict[nwp_source] = ImageEmbedding(
-                        318, nwp_sequence_len, self.nwp_encoders_dict[nwp_source].image_size_pixels
+                        image_embedding_dim,
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py:200]: Similarly can we rename this

File: pvnet/models/multimodal/multimodal.py [hunk 20]
```diff
@@ -241,22 +304,45 @@ def forward(self, x):
                 modes["pv"] = self.pv_encoder(x)
             else:
                 # Target is PV, so only take the history
-                pv_history = x[BatchKey.pv][:, : self.history_len_30].float()
-                modes["pv"] = self.pv_encoder(pv_history)
+                # Copy batch
+                x_tmp = x.copy()
+                x_tmp[BatchKey.pv] = x_tmp[BatchKey.pv][:, : self.history_len + 1]
```
Comment (dfulu) [inline pvnet/models/multimodal/multimodal.py:309]: Are you sure about this +1? That would give it the PV data at exactly t0 right?

File: pvnet/models/multimodal/multimodal.py [hunk 21]
```diff
@@ -176,22 +209,52 @@ def __init__(
             assert pv_history_minutes is not None
 
             self.pv_encoder = pv_encoder(
-                sequence_length=pv_history_minutes // 5 + 1,
```
Comment (jacobbieker) [inline pvnet/models/multimodal/multimodal.py:179]: @dfulu This is where I was getting the +1 used below. I can remove all the +1's, having them would include t0 in the inputs.

File: pvnet/models/multimodal/multimodal.py (hunk 20 above)
Comment (jacobbieker) [reply pvnet/models/multimodal/multimodal.py:309]: See my comment above, it should give the t0 to the input, but that I think is how the current one works for PV. I think there probably shouldn't be the +1, but didn't want to change the current behaviour.

Comment (peterdudfield) [reply pvnet/models/multimodal/multimodal.py:307]: The live data gets updated every 3 mins, with about a 10 minute delay. 
I think this is fine, and in production we can do a small bit of interpolation. We can always change this later

File: pvnet/models/multimodal/multimodal.py (hunk 21 above)
Comment (dfulu) [reply pvnet/models/multimodal/multimodal.py:179]: Ah okay, well then maybe ignore my previous comment for now and lets get this merged. Probably something we should come back to though

**Review by dfulu — APPROVED** [review]
Looks really good, thanks for all the work!
